src/pathfinding/grid_pathfinder.jl

```diff
@@ -0,0 +1,281 @@
+[`Path{D}`](@ref). This function usually does not need to be called explicitly, instead the use the provided [`set_target!`](@ref)
+and [`move_agent!`](@ref) functions.
+"""
+function find_path(pathfinder::AStar{D}, from::Dims{D}, to::Dims{D}) where {D}
+    grid = DefaultDict{Dims{D},GridCell}(GridCell(typemax(Int), typemax(Int), typemax(Int)))
+    parent = DefaultDict{Dims{D},Union{Nothing,Dims{D}}}(nothing)
+
+    open_list = MutableBinaryMinHeap{Tuple{Int,Dims{D}}}()
+    open_list_handles = Dict{Dims{D},Int64}()
+    closed_list = Set{Dims{D}}()
+
+    grid[from] = GridCell(0, delta_cost(pathfinder, from, to), pathfinder.admissibility)
+    push!(open_list, (grid[from].f, from))
+
+    while !isempty(open_list)
+        _, cur = pop!(open_list)
+        cur == to && break
+        push!(closed_list, cur)
+
+        nbors = get_neighbors(cur, pathfinder)
+        for nbor in Iterators.filter(n -> inbounds(n, pathfinder, closed_list), nbors)
+            new_g_cost = grid[cur].g + delta_cost(pathfinder, cur, nbor)
+            if new_g_cost < grid[nbor].g
+                parent[nbor] = cur
+                grid[nbor] = GridCell(
+                    new_g_cost,
+                    delta_cost(pathfinder, nbor, to),
+                    pathfinder.admissibility,
+                )
+                if haskey(open_list_handles, nbor)
+                    update!(open_list, open_list_handles[nbor], (grid[nbor].f, nbor))
+                else
+                    open_list_handles[nbor] = push!(open_list, (grid[nbor].f, nbor))
+                end
+            end
+        end
+    end
+
+    agent_path = Path{D}()
+    cur = to
+    while parent[cur] !== nothing
+        pushfirst!(agent_path, cur)
+        cur = parent[cur]
+    end
+    return agent_path
+end
+
+@inline get_neighbors(cur, pathfinder::AStar{D,true}) where {D} =
+    (mod1.(cur .+ β.I, pathfinder.grid_dims) for β in pathfinder.neighborhood)
+@inline get_neighbors(cur, pathfinder::AStar{D,false}) where {D} =
+    (cur .+ β.I for β in pathfinder.neighborhood)
+@inline inbounds(n, pathfinder, closed) =
+    all(1 .<= n .<= pathfinder.grid_dims) && pathfinder.walkable[n...] && n ∉ closed
+
+"""
+    set_target!(agent, pathfinder::AStar{D}, target::NTuple{D,Int})
+This calculates and stores the shortest path to move the agent from its current position to `target`
+using [`find_path`](@ref).
+"""
+function set_target!(agent, pathfinder::AStar{D}, target::Dims{D}) where {D}
+    pathfinder.agent_paths[agent.id] = find_path(pathfinder, agent.pos, target)
+end
+
+"""
+    move_agent!(agent, model::ABM{<:GridSpace{D}}, pathfinder::AStar{D})
+Moves the agent along the path to its target set by [`set_target!`](@ref). If the agent does
+not have a precalculated path, or the path is empty, nothing happens.
+"""
+function move_agent!(agent, model::ABM{<:GridSpace{D}}, pathfinder::AStar{D}) where {D}
+    (
+        get(pathfinder.agent_paths, agent.id, nothing) === nothing ||
```

Comment:
What's the advantage to using DefaultDict here? Couldn't we use a regular dict and use `haskey` in situations like this?

Reply:
Convenience, really. Since `DataStructures.jl` is a dependency anyway, I thought I'd use it. Now that you mention it, I guess it's worth profiling. I'll do it now

Reply:
I think using `Dict` is a hair faster
- `0` admissibility: `17.530 s (97244868 allocations: 11.96 GiB)`
- `0.3` admissbility: `3.483 s (18953562 allocations: 2.74 GiB)`
I'd say it's worth the change. This can scale up, and it's really only a couple more lines

Reply:
OK, great! I've got to stop working on this today, but it looks nearly done! Just some polish now.